File: examples/lin_algebra/qr.cpp
// Copyright 2023 The EA Authors.
// part of Elastic AI Search
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include <flare.h>
#include <cstdio>
#include <cstdlib>

using namespace fly;

int main(int argc, char* argv[]) {
    try {
        // Select a device and display flare info
        int device = argc > 1 ? atoi(argv[1]) : 0;
        fly::setDevice(device);
        fly::info();

        printf("Running QR InPlace\n");
        array in = randu(5, 8);
        fly_print(in);

        array qin = in.copy();

        array tau;
        qrInPlace(tau, qin);

        fly_print(qin);
        fly_print(tau);

        printf("Running QR with Q and R factorization\n");
        array q, r;
        qr(q, r, tau, in);

        fly_print(q);
        fly_print(r);
        fly_print(tau);

    } catch (fly::exception& e) {
        fprintf(stderr, "%s\n", e.what());
        throw;
    }

    return 0;
}

File: python/examples/lin_algebra/qr.py
#!/usr/bin/env python
##########################################################################
# Copyright 2023 The EA Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

import flare as fly

def main():
    try:
        fly.info()
        in_array = fly.randu(5,8)

        print("Running QR InPlace\n")
        q_in = in_array.copy()
        print(q_in)

        tau = fly.qr_inplace(q_in)

        print(q_in)
        print(tau)

        print("Running QR with Q and R factorization\n")
        q, r, tau = fly.qr(in_array)

        print(q)
        print(r)
        print(tau)

    except Exception as e:
        print("Error: ", str(e))

if __name__ == '__main__':
    main()
